File: C/15.命令模式/src/Stock.c
/**
 * @file Stock.c
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2020-08-09
 *
 * @copyright Copyright (c) 2020
 *
 */

#include "Stock.h"

#include <stdio.h>

#include "ctools.h"

struct Stock {
    const char* name;
    int quantity;
};

Stock* stock_create(void) {
    Stock* stock = (Stock*)calloc(1, sizeof(Stock));
    if (NULL == stock) return NULL;

    stock->name = "ABC";
    stock->quantity = 10;
    return stock;
}

void stock_destroy(Stock** stock) {
    if (NULL == stock || NULL == *stock) return;
    freep((void**)stock);
}

void stock_buy(Stock* stock) {
    if (NULL == stock) return;
    printf("Stock [ Name: %s, Quantity: %d ] bought\n", stock->name, stock->quantity);
}

void stock_sell(Stock* stock) {
    if (NULL == stock) return;
    printf("Stock [ Name: %s, Quantity: %d ] sold\n", stock->name, stock->quantity);
}

File: Objective-C/15.CommandPattern/Stock.m
//
//  Stock.m
//  15.CommandPattern
//
//  Created by Layne on 2020/11/5.
//

#import "Stock.h"

@implementation Stock {
    NSString* _name;
    int _quantity;
};

- (instancetype)init {
    self = [super init];
    if (self) {
        _name = @"ABC";
        _quantity = 10;
    }
    return self;
}

- (void)buy {
    NSLog(@"Stock [ Name: %@, Quantity: %d ] bought", _name, _quantity);
}

- (void)sell {
    NSLog(@"Stock [ Name: %@, Quantity: %d ] sold", _name, _quantity);
}

@end
